src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -903,15 +903,14 @@ private void WriteString(in DbRow row, Utf8JsonWriter writer)
             int numberOfRowsForMembers = 0;
             int numberOfRowsForValues = 0;
 
-            ref byte jsonStart = ref MemoryMarshal.GetReference(utf8JsonSpan);
-
             while (reader.Read())
             {
                 JsonTokenType tokenType = reader.TokenType;
 
-                int tokenStart = Unsafe.ByteOffset(
-                    ref jsonStart,
-                    ref MemoryMarshal.GetReference(reader.ValueSpan)).ToInt32();
+                // Since the input payload is contained within a Span, 
+                // token start index can never be larger than int.MaxValue (i.e. utf8JsonSpan.Length).
+                Debug.Assert(reader.TokenStartIndex <= int.MaxValue);
```

Comment:
@bartonjs, please review the changes related to JsonDocument

Reply:
Also, what should we do about the code that previously relied on `OriginalSpan`/`OriginalSequence`?
https://github.com/dotnet/corefx/blob/4e1b64e1ab06202e9aa451a0b320b7fd4ed11519/src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs#L424-L438

Reply:
> Also, what (if anything) should we do about the code that previously relied on OriginalSpan/OriginalSequence?

That code is to copy the values into new memory for the parsed JsonDocument, so it still needs OriginalSpan/OriginalSequence.  But I think this was before your last commit, so maybe it's already addressed?